mvcc/kvstore.go

```diff
@@ -160,6 +166,42 @@ func (s *store) Hash() (hash uint32, revision int64, err error) {
 	return h, s.currentRev, err
 }
 
+func (s *store) HashByRev(rev int64) (hash uint32, compactRev int64, err error) {
+	for {
+		s.mu.RLock()
+		compactMainRev := s.compactMainRev
+		currentRev := s.currentRev
```

Comment:
The revision at the time of starting the hash. It'll matter more once it's nonblocking. Consider:

Start Hash(rev=10); currentRev=100
Compact(rev=50); currentRev=100
Put; currentRev=101
Finish Hash; currentRev=101

If Hash returns 101 as its header revision instead of 100, then etcd is saying the Hash would have worked at rev=101. This cannot be true since the keys for the hash at rev=10 have been compacted away by rev=101.

This is similar to the reasoning for Range's header revisions. If I request a Get with rev=0, then I'd expect the same KV values (or a compaction error) if I make a request later using WithRev(getResp.Header.Revision). This is guaranteed if the header is populated with the revision at the time of starting the read transaction, but not if it's using the revision after the transaction completes.